ledger/src/token_balances.rs

```diff
@@ -10,34 +11,41 @@ use {
         token_balances::TransactionTokenBalances, TransactionTokenBalance,
     },
     spl_token_2022::{
-        extension::StateWithExtensions,
+        extension::{
+            interest_bearing_mint::InterestBearingConfig, BaseStateWithExtensions,
+            StateWithExtensions,
+        },
         state::{Account as TokenAccount, Mint},
     },
     std::collections::HashMap,
 };
 
-fn get_mint_decimals(bank: &Bank, mint: &Pubkey) -> Option<u8> {
+fn get_mint_additional_data(bank: &Bank, mint: &Pubkey) -> Option<SplTokenAdditionalData> {
     if mint == &spl_token::native_mint::id() {
-        Some(spl_token::native_mint::DECIMALS)
+        Some(SplTokenAdditionalData::with_decimals(
+            spl_token::native_mint::DECIMALS,
+        ))
     } else {
         let mint_account = bank.get_account(mint)?;
-
         if !is_known_spl_token_id(mint_account.owner()) {
             return None;
         }
-
-        let decimals = StateWithExtensions::<Mint>::unpack(mint_account.data())
-            .map(|mint| mint.base.decimals)
-            .ok()?;
-
-        Some(decimals)
+        let mint_state = StateWithExtensions::<Mint>::unpack(mint_account.data()).ok()?;
+        let interest_bearing_config = mint_state
+            .get_extension::<InterestBearingConfig>()
+            .map(|x| (*x, bank.clock().unix_timestamp))
+            .ok();
```

Comment:
Hmm, I'm wondering whether we want to do this, at least behind the scenes. I'm not fully decided, but it seems a little weird to me to have token pre- and post-balances on a transaction reflect something other than the actual quantity in the account. The RPC changes seem less bad, because a parse token account includes information about extensions. What are your thoughts?

Reply:
My thinking was that `UiTokenAmount` contains `amount`, `decimals`, and `ui_amount_string`, so callers can use the relevant information for them. If they care about raw amounts, `amount` is available, but if they want to show `ui_amount_string`, that's there too. If they want to use the old `ui_string_amount`, they can recover that using `amount` and `decimals`.

On the flipside, there is also the issue that there will be an abrupt disconnect for this field on old transaction data -- as soon as an RPC node starts using this code, it will report different amounts, as noted at https://github.com/anza-xyz/agave/pull/1549#issuecomment-2143113186

I was aiming for consistency everywhere, but you're right that this change will be destructive. Since instruction data doesn't report the correct UI amount, it might be reasonable to say that transactions shouldn't either.

All that to say, I think I'm OK dropping this bit, but let me sleep on it.

Reply:
Dropped it in [aaa79b4](https://github.com/anza-xyz/agave/pull/1549/commits/aaa79b49d737e610946f1fd3185c6885f9bf642d)